accounts/checkers.go

```diff
@@ -0,0 +1,407 @@
+			r.Channels = []*lnrpc.ChannelCloseSummary{}
+
+			return r, nil
+		},
+	)
+
+	WalletBalanceEmptyRewriter = rpcmiddleware.NewResponseRewriter(
+		&lnrpc.WalletBalanceRequest{},
+		&lnrpc.WalletBalanceResponse{},
+		func(t *lnrpc.WalletBalanceResponse) (proto.Message, error) {
+			t.ConfirmedBalance = 0
+			t.TotalBalance = 0
+			t.UnconfirmedBalance = 0
+			for acctName := range t.AccountBalance {
+				t.AccountBalance[acctName].ConfirmedBalance = 0
+				t.AccountBalance[acctName].UnconfirmedBalance = 0
+			}
+
+			return t, nil
+		},
+	)
+
+	GetTransactionsEmptyRewriter = rpcmiddleware.NewResponseRewriter(
+		&lnrpc.GetTransactionsRequest{},
+		&lnrpc.TransactionDetails{},
+		func(r *lnrpc.TransactionDetails) (proto.Message, error) {
+			r.Transactions = make([]*lnrpc.Transaction, 0)
+
+			return r, nil
+		},
+	)
+
+	ListPeersEmptyRewriter = rpcmiddleware.NewResponseRewriter(
+		&lnrpc.ListPeersRequest{}, &lnrpc.ListPeersResponse{},
+		func(r *lnrpc.ListPeersResponse) (proto.Message, error) {
+			r.Peers = []*lnrpc.Peer{}
+
+			return r, nil
+		},
+	)
+)
+
+func (s *Service) generateCheckers(
+	acct *OffChainBalanceAccount) map[string]rpcmiddleware.RoundTripChecker {
+
+	return map[string]rpcmiddleware.RoundTripChecker{
+		// Invoices:
+		"/lnrpc.Lightning/AddInvoice": rpcmiddleware.NewResponseRewriter(
+			&lnrpc.Invoice{},
+			&lnrpc.AddInvoiceResponse{},
+			func(t *lnrpc.AddInvoiceResponse) (proto.Message,
+				error) {
+
+				hash, err := lntypes.MakeHash(t.RHash)
+				if err != nil {
+					return nil, fmt.Errorf("error parsing "+
+						"invoice hash: %v", err)
+				}
+
+				return nil, s.Store.associateInvoice(
+					acct.ID, hash,
+				)
+			},
+		),
+		"/lnrpc.Lightning/ListInvoices": rpcmiddleware.NewResponseRewriter(
+			&lnrpc.ListInvoiceRequest{},
+			&lnrpc.ListInvoiceResponse{},
+			func(t *lnrpc.ListInvoiceResponse) (proto.Message,
+				error) {
+
+				filteredInvoices := make(
+					[]*lnrpc.Invoice, 0, len(t.Invoices),
+				)
+				for _, invoice := range t.Invoices {
+					hash, err := lntypes.MakeHash(
+						invoice.RHash,
+					)
+					if err != nil {
+						return nil, err
+					}
+
+					if _, ok := acct.Invoices[hash]; ok {
+						filteredInvoices = append(
+							filteredInvoices,
+							invoice,
+						)
+					}
+				}
+
+				t.Invoices = filteredInvoices
+				return t, nil
+			},
+		),
+		"/lnrpc.Lightning/LookupInvoice": rpcmiddleware.NewRequestChecker(
+			&lnrpc.PaymentHash{},
+			&lnrpc.Invoice{},
+			func(t *lnrpc.PaymentHash) error {
+				hash, err := lntypes.MakeHash(t.RHash)
+				if err != nil {
+					return err
+				}
+				_, ok := acct.Invoices[hash]
+				if !ok {
+					return fmt.Errorf("invoice does not " +
+						"belong to this account")
+				}
+
+				return nil
+			},
+		),
+
+		// Payments:
+		"/lnrpc.Lightning/SendPayment": rpcmiddleware.NewRequestChecker(
+			&lnrpc.SendRequest{},
+			&lnrpc.SendResponse{},
+			func(r *lnrpc.SendRequest) error {
+				return s.checkSend(
```

Comment:
The problem is that we can only map an incoming payment to an account if we see the invoice being created with the macaroon. We don't know that a keysend payment is for a specific account. Unless it would encode that information somehow, which is something very custom that each wallet would need to implement... I guess we would need to use the HTLC interceptor (and a bunch of other tricks) to generate a custom fake destination node pubkey and then show that to the user so someone else could keysend them.
Definitely out of scope of this PR.

Reply:
ah yes, good point 👍